File: iOS/Features/Sources/Features/Models/Item.swift
import Foundation

// MARK: - Supporting enums

public enum ItemLocation: String, Codable, CaseIterable, Equatable, Sendable {
    case fridge
    case pantry
    case freezer
    case custom

    public var displayName: String {
        switch self {
        case .fridge:  return "Fridge"
        case .pantry:  return "Pantry"
        case .freezer: return "Freezer"
        case .custom:  return "Custom"
        }
    }

    public var systemImage: String {
        switch self {
        case .fridge:  return "refrigerator"
        case .pantry:  return "cabinet"
        case .freezer: return "snowflake"
        case .custom:  return "ellipsis.circle"
        }
    }
}

public enum Freshness: Equatable, Sendable {
    case fresh
    case expiring
    case expired
}

// MARK: - Item

public struct Item: Identifiable, Codable, Equatable, Sendable {
    public let id: UUID
    public var name: String
    public var quantity: String?
    public var location: ItemLocation
    public var locationCustom: String?
    public var bestBeforeDate: Date
    public var addedDate: Date
    public var remindDaysBefore: Int
    public var remindOnDay: Bool
    public var isUsed: Bool

    public init(
        id: UUID = UUID(),
        name: String,
        quantity: String? = nil,
        location: ItemLocation = .fridge,
        locationCustom: String? = nil,
        bestBeforeDate: Date,
        addedDate: Date = Date(),
        remindDaysBefore: Int = 3,
        remindOnDay: Bool = false,
        isUsed: Bool = false
    ) {
        self.id = id
        self.name = name
        self.quantity = quantity
        self.location = location
        self.locationCustom = locationCustom
        self.bestBeforeDate = bestBeforeDate
        self.addedDate = addedDate
        self.remindDaysBefore = remindDaysBefore
        self.remindOnDay = remindOnDay
        self.isUsed = isUsed
    }

    // MARK: - Computed

    public var daysLeft: Int {
        Calendar.current.dateComponents([.day], from: .now, to: bestBeforeDate).day ?? 0
    }

    public var remindOn: Date {
        Calendar.current.date(byAdding: .day, value: -remindDaysBefore, to: bestBeforeDate) ?? bestBeforeDate
    }

    public var freshness: Freshness {
        if daysLeft < 0 { return .expired }
        if daysLeft <= remindDaysBefore { return .expiring }
        return .fresh
    }

    public var locationLabel: String {
        location == .custom ? (locationCustom ?? location.displayName) : location.displayName
    }
}

// MARK: - ItemIn (API request body)

public struct ItemIn: Codable, Equatable, Sendable {
    public var name: String
    public var quantity: String?
    public var location: ItemLocation
    public var locationCustom: String?
    public var bestBeforeDate: Date
    public var remindDaysBefore: Int
    public var remindOnDay: Bool

    public init(
        name: String,
        quantity: String? = nil,
        location: ItemLocation = .fridge,
        locationCustom: String? = nil,
        bestBeforeDate: Date,
        remindDaysBefore: Int = 3,
        remindOnDay: Bool = false
    ) {
        self.name = name
        self.quantity = quantity
        self.location = location
        self.locationCustom = locationCustom
        self.bestBeforeDate = bestBeforeDate
        self.remindDaysBefore = remindDaysBefore
        self.remindOnDay = remindOnDay
    }
}

File: Backend/app/schemas/item.py
"""Pydantic schemas for the items (expiry-date tracker) endpoints."""
from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, computed_field, field_validator

VALID_LOCATIONS = frozenset({"fridge", "pantry", "freezer", "custom"})


class ItemIn(BaseModel):
    name: str
    quantity: Optional[str] = None
    location: Optional[str] = None
    location_custom: Optional[str] = None
    best_before_date: date
    added_date: Optional[date] = None
    remind_days_before: int = 3
    remind_on_day: bool = False

    @field_validator("location")
    @classmethod
    def validate_location(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in VALID_LOCATIONS:
            raise ValueError(f"location must be one of {sorted(VALID_LOCATIONS)}")
        return v


class ItemUpdate(BaseModel):
    name: Optional[str] = None
    quantity: Optional[str] = None
    location: Optional[str] = None
    location_custom: Optional[str] = None
    best_before_date: Optional[date] = None
    remind_days_before: Optional[int] = None
    remind_on_day: Optional[bool] = None
    is_used: Optional[bool] = None

    @field_validator("location")
    @classmethod
    def validate_location(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in VALID_LOCATIONS:
            raise ValueError(f"location must be one of {sorted(VALID_LOCATIONS)}")
        return v


class ItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    name: str
    quantity: Optional[str] = None
    location: Optional[str] = None
    location_custom: Optional[str] = None
    best_before_date: date
    added_date: date
    remind_days_before: int
    remind_on_day: bool
    is_used: bool
    created_at: datetime

    @computed_field  # type: ignore[prop-decorator]
    @property
    def days_left(self) -> int:
        """Signed days until best_before_date from today. Negative = already expired."""
        from datetime import date as date_type  # local import avoids circular refs

        return (self.best_before_date - date_type.today()).days

    @computed_field  # type: ignore[prop-decorator]
    @property
    def remind_on(self) -> date:
        """The calendar date on which the advance reminder fires."""
        from datetime import timedelta

        return self.best_before_date - timedelta(days=self.remind_days_before)

    @computed_field  # type: ignore[prop-decorator]
    @property
    def freshness(self) -> Literal["fresh", "expiring", "expired"]:
        """
        'expired'  — best_before_date is in the past
        'expiring' — within remind_days_before days (inclusive)
        'fresh'    — more than remind_days_before days away
        """
        d = self.days_left
        if d < 0:
            return "expired"
        if d <= self.remind_days_before:
            return "expiring"
        return "fresh"
